site/bps/BP-38-bookie-endpoint-discovery.md

```diff
@@ -0,0 +1,149 @@
+        },
+        {
+         "name": "bookie-http-server",
+         "hostname": "localhost",
+         "port": 8080,
+         "protocol": "http",
+         "auth": [],
+         "extensions": []
+        }
+    ],
+    properties {
+        "property": "value",
+        "property2": "value2"
+    }
+}
+```
+Such information will be stored inside the '/REGISTRATIONPATH/available' znode (or /REGISTRATIONPATH/available/readonly' in case of readonly bookie), these paths are the same used in 4.10, but in 4.10 we are writing empty znodes.
+
+The rationale around this choice is that the client is already using these znodes in order to discover available bookies services.
+
+It is possible that such endpoint information change during the lifetime of a Bookie, like after rebooting a machine and the machine gets a new network address.
+
+It is out of the scope of this proposal to change semantics of ledger metadata, in which we  are currently storing directly the network address of the bookies, but with this proposal we are preparing the way to have an indirection and separate the concept of Bookie ID from the Network address.
+
+**Endpoint structure**
+- **name**: this is an id for the service type, like 'bookie'
+- **hostname**: the hostname (or a raw IP address)
+- **port**: the TCP port (int)
+- **protocol**: the type of service (bookie-rpc, http, https)
+- **auth**: supported authentication types (tls-auth, sasl-gssapi...)
+- **extensions**: protocol extensions, like 'starttls'
+
+**Well known additional properties**
+- **metrics.type**: the type of metrics, for instance 'prometheus' for Prometheus.io metrics
+- **autorecovery.enabled**: 'true' case of the presence of the auto recovery daemon
+
+In the future we could also leverage this information from the PlacementPolicy, in order to select only bookies that publish a TLS endpoint or a particular authentication mechanism.
+The GRPC service would use this mechanism as well.
+
+#### Which kind of properties should be stored in BookieServiceInfo ?
+
+We should store here only the minimal set of information useful to reach the bookie in an efficient way.
+So we are not storing on the Metadata Service information about the internal state of the server: if you know the address of an HTTP endpoint you can use the REST API to query the Bookie for its state.
+
+These properties may change during the lifetime of a Bookie, think about a configuration (change auth type, enable TLS, change network address) or a dynamic assigned DNS name.
+
+It is better not to expose the version of the Bookie, if the client wants to use particular features of the Bookie this should be implemented on the protocol itself, not just by using the version. The version of the Bookie could be exposed on the HTTP endpoint.
+
+### Proposed Changes
+
+See the 'Public interfaces' section.
+
+### Compatibility, Deprecation, and Migration Plan
+
+The proposed change will be compatible with existing clients up to 4.10.x.
```

Comment:
@eolivelli Where is the break? 4.10.x is a very recent version.

Reply:
There is no break, we will be compatible with every version of BK, up to 4.3 (personally I don't know how it worked before 4.3).
Maybe the English is not good ?
I meant to say that this change will be compatible with every client from xxx (4.3?) up to 4.10

Reply:
It's a little unclear. The fact that you state "up to 4.10.x", implies that anything after 4.10.x is _not_ compatible. So then I took it to mean the inverse, 4.10.x onwards. This should state that it's compatible with all clients from 4.3 onwards. Maybe a note on what the existing clients will do (i.e. just read the list of znodes, which doesn't change). I think the last break in this area was to use hostnames, and I don't recall that even broke anything.

Reply:
@ivankelly I have removed the reference to the version.
Now it reads 'The proposed change will be compatible with existing clients'

Reply:
Sure, sgtm.